src/ripple/app/tx/impl/CFTokenIssuanceCreate.cpp

```diff
@@ -0,0 +1,127 @@
+//------------------------------------------------------------------------------
+/*
+  This file is part of rippled: https://github.com/ripple/rippled
+  Copyright (c) 2023 Ripple Labs Inc.
+
+  Permission to use, copy, modify, and/or distribute this software for any
+  purpose  with  or without fee is hereby granted, provided that the above
+  copyright notice and this permission notice appear in all copies.
+
+  THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+  WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+  MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+  ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+  WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+  ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/app/tx/impl/CFTokenIssuanceCreate.h>
+#include <ripple/ledger/View.h>
+#include <ripple/protocol/Feature.h>
+#include <ripple/protocol/TxFlags.h>
+#include <ripple/protocol/st.h>
+
+namespace ripple {
+
+NotTEC
+CFTokenIssuanceCreate::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureCFTokensV1))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    if (ctx.tx.getFlags() & tfCFTokenIssuanceCreateMask)
+        return temINVALID_FLAG;
+
+    if (auto const fee = ctx.tx[~sfTransferFee])
+    {
+        if (fee > maxTransferFee)
+            return temBAD_CFTOKEN_TRANSFER_FEE;
+
+        // If a non-zero TransferFee is set then the tfTransferable flag
+        // must also be set.
+        if (fee > 0u && !ctx.tx.isFlag(tfTransferable))
+            return temMALFORMED;
+    }
+
+    // TODO length?
+    // if (auto const code = ctx.tx[~sfAssetCode]; code.length() >
+    // maxAssetCodeLength)
+    //     return temMALFORMED;
+
+    if (auto const metadata = ctx.tx[~sfCFTokenMetadata])
+    {
+        if (metadata->length() == 0 ||
+            metadata->length() > maxCFTokenMetadataLength)
+            return temMALFORMED;
+    }
+
+    return preflight2(ctx);
+}
+
+TER
+CFTokenIssuanceCreate::preclaim(PreclaimContext const& ctx)
+{
+    // TODO - check number of CFTs? not sure why we need to limit it
```

Comment:
agreed - i left this here because i'm not sure what the rationale was in the original spec behind limiting the number of issuances an account can have. i don't see any reason to. considering leaving this TODO here for now for this PR and not implementing a limit

Reply:
> i'm not sure what the rationale was in the original spec behind limiting the number of issuances an account can have

CFTs were forged out of the ledger instability events we all observed in late 2021. In response to that, we wanted to create something that (1) is much smaller on ledger than trustlines/RippleState objects, and (2) efficient for a non-issuer to hold _many_ of. 

I think the `CFTokenPage` accomplishes that for token _**holders**_.

Now, we could have specified some similar "more efficient storage mechanism" (i.e., paging) for token **_Issuers_** as well, but we observed that it's likely bad practice to issue multiple tokens from a single account (see section [1.1.2.2 in the Assumptions](https://github.com/XRPLF/XRPL-Standards/tree/master/XLS-0033d-CFTs#112-assumptions) section of the spec).

I personally would consider limiting token issuance to only 1 per issuing account (to properly accommodate global freeze); however, 32 seemed like a good balance between `1` (which limits issuers) and `many` -- though in truth this is an  arbitrary number.

> So from my perspective we have two choices:

Would a potential third choice be to simply introduce a counter (on the account I guess) that increments and decrements whenever an issuance is created or destroyed? (Not sure if that's a good idea or not, but we do something similar with `OutstandingAmount` (i.e., this value changes when new tokens are "issued").

Reply:
Yes.  Introducing a counter on the `AccountRoot` is a viable choice.  And if I had been more fair in my initial comment I should have included that third choice.

Introducing a counter is, in fact, how we handle minting of`NFToken`s.  We had such a good time with that field that we now have four (optional) fields on the `AccountRoot` that are only ever used by issuers of NFTs.  Once the first such field was added then the other three just kind of grew there naturally.  It's a bit like seeding a crystal.

Personlly, when I look at that (NFToken) change now, I'm not fond of it.  In my opinion very few users of the XRP Ledger are likely to mint NFTs.  They probably should not be such a dominant presence in the `AccountRoot`.  But here we are.  My feelings are similar with CFT issuers: we should expect to see very few on the ledger.

That said, I'm 100% aligned with your point.  There are at least three different ways to address this requirement.  There may be more.  It's worth looking at all the options we identify and deciding which we like best.

Reply:
Circling back to this thread...

> The theoretical maximum number of directory nodes to walk is `32 x 262144 = 8,388,608 nodes` just to see if a CFTokenIssuance can be added to an account.

I'd like to probe a bit deeper into this assertion.

The intention in the current CFT spec is that any `CFTokenIssuance` would be tracked in an OwnerDirectory with a `space key` (currently 0x7E) dedicated to CFTs (Note: the implementation will need to ensure that only CFT issuances get tracked by this type of directory).

So, to compute `how many CFTs does an issuer currently have`, I had the following "algorithm" in mind:

1. Compute the `RootIndex` for an issuer's CFT OwnerDirectory (see specific process [here](https://xrpl.org/directorynode.html#directory-id-formats) which hashes the `space-key` and the issuer's `AccountId`).
2. Using that key, lookup the first page (i.e., the "root") of that OwnerDirectory.
3. Count the number of entries in the root directory's `Indexes` field ([this example](https://xrpl.org/directorynode.html#example-directorynode-json), while not CFTs, has 2).

This works because:

a. Any OwnerDirectory page can store _up to_ 32 entries.
b. The spec limits the number of CFT issuances (per-issuer) to 32.
c. Therefore, no issuer account should ever have more than one page in the CFT OwnerDirectory.

This should mean it will never be necessary to find any `previous` or `next` directory pages, which means we can trivially count the number of entries in the root directory's `Indexes` field with only one lookup and some array counting.

Finally, to solve for `Has an issuer already issued a particular CFT`, we can [compute the CFTokenIssuanceId](https://github.com/XRPLF/XRPL-Standards/tree/master/XLS-0033d-CFTs#12111-cftokenissuance-ledger-identifier), and then check the `Indexes` field per the above.

OK, with all that:

1. @scottschurr does what I wrote change your opinion around efficiency?
2. Do my assumptions make sense, or am I misunderstanding something (e.g., WRT DirectoryNodes or otherwise)?
3. Does the CFT specification need to be adjusted to reflect what I wrote above? (e.g., [1.2.1.1.1](https://github.com/XRPLF/XRPL-Standards/tree/master/XLS-0033d-CFTs#12111-fields) and [1.2.1.1.2](https://github.com/XRPLF/XRPL-Standards/tree/master/XLS-0033d-CFTs#12112-fields) seem to cover the space key and owner directory nuance -- but maybe that can be improved in the spec?).

Reply:
Hi @sappenin, I think what you're describing is very much what I was waving my hand at here:

>   2. Store the `CFTokenIssuances` in a way that is conducive to easy counting.
>
> Implementing choice 2 would also be easy. One way would be to store the `CFTokenIssuances` in a sorted array that is owned by the account. Not so different from how we handle SignerLists.

Since there is an upper limit of 32 issuances there's no point in having a directory.  We can just store the entries in an array (the way a `SignerList` works).  And, yes, the `CFTokenIssuances` array would be indexed using a hash of the `space-key` for `CFTokenIssuances` and the issuer's `AccountID`.

So I guess I misunderstood the original spec.  But what you're describing is not an `OwnerDirectory` at all.  That concept is already spoken for and all `AccountRoot`s have at most one.  If I'm understanding you correctly what you're describing is more akin to the `SignerList` that an account may optionally own.

And to answer question 3: Yes, I think this approach can be taken very efficiently (modulo a different problem I'll describe below)

Am I making sense?  Or did I misunderstand what you're saying?  If you'd like me to go into more detail regarding the `SignerList` I'm happy to do so.

That said, there may be a downside to the approach I'm describing.  Whether it _is_ a problem depends on which data is tracked in the `CFTokenIssuances` array.  Let's just say that each `CFTokenIssuance` tracks how much of that issuance is held by other accounts, and it tracks that amount somewhere in the issuance's entry in the `CFTokenIssuances` array.

If an issuance has a fee associated with it, the amount of that issuance in the ledger will go down by some small amount every time that issuance changes hands.  So the value _tracking_ that on-ledger amount would get a little bit smaller every time the issuance changes hands.  The result is that some value in the `CFTokenIssuance` array changes slightly every time any of the (up to) 32 issuances in that array changes hands.  That seems benign...

The problem is the size of the metadata that is produced when the contents of an array changes on ledger.  The current behavior is that the contents of the entire modified array is saved in the metadata any time anything in the array changes.  We don't usually think of lots of metadata as a problem, but every bit of metadata for every transaction is stored by a full history server (and/or clio).  We don't want to be needlessly producing a lot of metadata.

`SignerList`s change very seldom, and when they are changed, an entire new `SignerList` is specified.  So what I've just described is not a problem for `SignerList`s.  But it _is_ a problem in `NFTokenPage`s.  So we know the problem is real.  There are different ways it could be worked around.  But, if the `CFTokenIssuances` array approach is used it will be worthwhile to accommodate this concern in the design.

Sorry for the side rant.  But I thought it would be an important consideration for future design work in this direction.

Reply:
Thanks for your responses here @scottschurr - they're really helpful! I'm only now getting a chance to respond, so hopefully my delays aren't a problem. Some more thoughts from me below.

> But what you're describing is not an OwnerDirectory at all.

I was thinking we would create a new type of [Directory](https://xrpl.org/directorynode.html#directorynode) -- something _like_ an `OwnerDirectory` -- but new, for CFTs and limited to only a single page (to make counting easy).

The advantage of this approach is that -- because we only store object hashes in an index in this new type of directory -- it doesn't suffer from the metadata problem you outlined (e.g., when one `CFTokenIssuance` changes, the metadata for all other issuances is not reflected).

I'm open to reasons why a new type of Directory might be undesirable. Keep reading for why your SignerList proposal is sub-optimal for me.

> The problem is the size of the metadata that is produced when the contents of an array changes on ledger. 

Yes, this could be a reasonably large problem depending on the number of "issue" transactions any issuer does -- I expect there to be a fair amount of issuance and redemption. Plus, any escrows will affect the `CFTokenIssuance` object, plus other transactions.

In my opinion, this disqualifies the "Store CFTIssuance like a SignerList" as the best option because it will produce a lot of extra metadata (though perhaps as tradeoffs go, there's some reason to like the "like a SignerList" idea and prefer it, despite the metadata?)

> Since there is an upper limit of 32 issuances there's no point in having a directory...

I'm not convinced here -- but is there a mechanism that you can suggest that doesn't suffer from the metadata problem? The only thing I can think of is my idea above about a new type of Directory.

> SignerLists change very seldom, and when they are changed, an entire new SignerList is specified. So what I've just described is not a problem for SignerLists. But it is a problem in NFTokenPages. So we know the problem is real.

I'm glad you pointed this out -- and I'm going to change the subject very subtly here -- from `CFTokenIssuance` (owned by issuers) to `CFTokenPages` (owned by token holders). 

The fact that `NFTokenPages` produce a lot of metadata is something I wasn't aware of, and hadn't thought of until now. But in the CFT spec, we used that design for `NFTokenPages` as the storage mechanism for CFT Token _holders_ (on the assumption that we want users to potentially hold _many_ tokens, if desired...just like NFTs).

In other words, if someone holds a lot of CFTs in a `CFTokenPage`, and makes a payment with a CFT, that _might_ generate a lot of metadata. Luckily, there are only four field in a Token, so the problem might not be as bad as for NFTs.

But in any event, do you have any thoughts/ideas about how to potentially solve metadata problem, whether for NFTs or for CFTs (again, in this last bit I'm speaking about CFToken holders, not issuers).

Reply:
Thanks for the response @sappenin.  It feels like we have split the topic into two parts.  For now I'd like to stick to the original thread (CFT issuance directories).  I'm glad you pointed out the metadata problems with `CFTokenPage`s.  That should have occurred to me earlier, and I'm glad you caught it.

At this point I think we are pretty well aligned on what the implementation ought to look like for CFT issuances in the situation where we allow a maximum of 32 to be owned by a single issuer.  We may not be agreeing on the terminology, but I think that's a smaller issue.

I'll describe what I think we're agreeing to:

1. An issuer has one (and only one) CFT "directory" page.  This "directory" contains an array of indexes to CFT issuance nodes.  It's primary reason for being it to track all of the CFT issuances owned by a single issuer.  As a side effect of its existence, it's easy to track the number of issuances from a single issuer, and limit them to 32.

2. Each separate issuance from an issuer also has its own node in the ledger.  This is where all of the accounting for that issuance takes place.  It doesn't need a back pointer to the CFT "directory" page, because there will always be at most one "directory" page.

This design is justified by the improved efficiency of the metadata when there is more than one CFT issuance from a single issuer.  It's worth considering that the design is certainly less efficient if most issuers issue just a single currency.  But assuming a significant percentage of issuers want to issue multiple currencies, then this seems like a good design.

On the terminology front, I'm still hesitant to call the single node that has an array of indexes to CFT issuances a "directory".  I see where you are coming from.  But I've spent a lot of time with rippled implementation.  There are specific library functions that are used to access directories, whether they are owner directories, book directories, or NFToken directories.  I don't think any of those library functions apply in this case.  That's largely because CFT issuance has exactly one or zero nodes to manage.  Everything else called a directory in the rippled code base manages multiple nodes in one way or another. 

 So my inclination would be to call the node that holds an array of indexes to CFT issuances a `CFTOwnedIssuancesList` or something like that.  To me that makes it clear that we're not dealing with linked lists of nodes and the like.  But that's just my opinion.

Reply:
good stuff, but all of this misses the key point IMO - ***why are we seeking to limit issuances per account***? Since I haven't heard a good answer to that thus far, I'm sticking with this implementation: you can have as many issuances as you like. There is no limit per account.

Reply:
happy to revisit in a separate, later PR if there is actually some reason to have a limit.  the implementations you each suggest both seem reasonable, but the requirement does not to me.

Reply:
i've incorporated this change to the spec in my open PR with-regard-to this feature

Reply:
> For now I'd like to stick to the original thread (CFT issuance directories)

@scottschurr I agree, good suggestion.

> At this point I think we are pretty well aligned

I agree. I thought all of your suggestions sound reasonable.

> It's worth considering that the design is certainly less efficient if most issuers issue just a single currency. But assuming a significant percentage of issuers want to issue multiple currencies, then this seems like a good design.

Can you say a bit more here? I personally expect a typical issuer to only issue one CFT, with some issuers issuing a handful (basically like we see in XRPL mainnet now). 

I'm curious what you would change in your proposed design if:

1. Issuers were limited to 1 CFT (I presume this would be just, "store it the owner directory"?)
2. Issuers were limited to 5 CFTs


> On the terminology front, I'm still hesitant to call the single node that has an array of indexes to CFT issuances a "directory". my inclination would be to call the node that holds an array of indexes to CFT issuances a `CFTOwnedIssuancesList` or something like that.

I agree. Perhaps even `CFTOwnedIssuances` could work (not sure if the prefix `List` is salient for you).

--

From my perspective, the only real sticking point (as you both identified @scottschurr and @ledhed2222) is: Should we limit the number of issuances? (I'm happy to move this over to [here](https://github.com/XRPLF/XRPL-Standards/pull/123)).

Reply:
> I'm curious what you would change in your proposed design if:
>
> 1. Issuers were limited to 1 CFT (I presume this would be just, "store it the owner directory"?)
>
> 2. Issuers were limited to 5 CFTs

As you already figured out, the magic threshold is between one and more than one.  If an issuer is limited to at most one CFT issuance then a couple of things get simpler:

1. The index for the issuance object in the ledger is just the hash of the space ID with the account ID of the issuer.

2. We no longer need anything that looks like a directory to track owned issuances.  Since an issuer has at most one issuance we can check the ledger for the index of that issuance to figure out whether they have exceeded the maximum.

The efficiency I mentioned earlier is just the loss of the directory-like thing and simpler indexing.

By the time we allow three CFT issuances from a single account then we want to shift over to the directory-like design currently being discussed.  I haven't thought through a design where we allow at most two, but that seems like a silly design.

The real heart of the efficiency is that, with the design under discussion, all issuers will carry a directory-like node.  That includes all issuers of only one currency.  So most of the directory-like nodes will have only one entry.  Those issuers don't _need_ a directory-like node, but they carry one anyway.

If most issuers issue only one currency we are paying the price to allow all of them to issue up to 32 CFTs.

That said, I want the design to be uniform.  If we allow any accounts to issue more than one currency, then all accounts that issue CFTs should own a directory-like node (in my opinion).

> I agree. Perhaps even `CFTOwnedIssuances` could work (not sure if the prefix List is salient for you).

This is a better name.  The `List` is unimportant.

Reply:
@scottschurr Great points - my current opinion is: (1) I don't expect there to be very many _issuers_ in the general case (maybe a few hundreds or couple thousand? [Famous last words](https://geekhistory.com/content/urban-legend-i-think-there-world-market-maybe-five-computers)). 

Anyway, If my expectations proves true, I don't think there's much worry between "allow 1 issuance per account to save some space" vs. "allow 32 such that we need a `CFTOwnedIssuances` object" -- basically because I don't expect _that_ many issuers.

So, for now I feel good about introducing `CFTOwnedIssuances` (but weakly held opinion. We can move this discussion over to [the spec PR](https://github.com/XRPLF/XRPL-Standards/pull/123) that @ledhed2222 made so we can unblock this code PR).

@ledhed2222 if you want to resolve this discussion, feel free -- @scottschurr and I can continue over in the other repo.

Reply:
we have agreed not to limit the number of cfts/account (for now)